compiler/rustc_trait_selection/src/traits/error_reporting/type_err_ctxt_ext.rs

```diff
@@ -639,6 +639,32 @@ impl<'tcx> TypeErrCtxt<'_, 'tcx> {
                             err.span_label(span, explanation);
                         }
 
+                        if Some(trait_ref.def_id()) == self.tcx.lang_items().sized_trait() {
+                            if !self.tcx.object_safety_violations(trait_ref.def_id()).is_empty() {
+                                match (
+                                    obligation.cause.code(),
+                                    trait_ref.skip_binder().self_ty().kind(),
+                                ) {
+                                    (
+                                        ObligationCauseCode::WhereClauseInExpr(..),
+                                        ty::Dynamic(..),
+                                    ) => {
+                                        return err.delay_as_bug();
+                                    }
+                                    (ObligationCauseCode::SizedCallReturnType(did), _) => {
```

Comment:
I'd like to note that these checks are like super specific / not really general. Cost in complexity vs. benefit.